src/execution_plan/ops/op_merge.c

```diff
@@ -185,18 +186,26 @@ static OpResult MergeInit(OpBase *opBase) {
 
 static Record _handoff(OpMerge *op) {
 	Record r = NULL;
-	if(array_len(op->output_records)) r = array_pop(op->output_records);
+	if(op->matched_records) {
+		if(array_len(op->matched_records)) {
+			r = array_pop(op->matched_records);
+		} else if(op->created_records) {
+			if(array_len(op->created_records)) {
+				r = array_pop(op->created_records);
+			}
+		}
+	}
```

Comment:
Separating `matched_records` and `created_records` (as opposed to just having `output_records` like before) seems like a more complicated than necessary to me. I guess the point is to invoke the create stream before making `ON MATCH` updates, so as to not muddle which writer is last? We could still just store the number of records to apply `ON MATCH` updates to prior to line 271, then `_UpdateProperties` for that many records.

Reply:
nice
done